src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,448 @@
+    metrics = get_metrics()
+    task = get_task()
+
+    strings = set()
+    parsed_payloads_by_offset = {
+        msg.offset: json.loads(msg.payload.value.decode("utf-8"), use_rapid_json=True)
+        for msg in outer_message.payload
+    }
+
+    for message in parsed_payloads_by_offset.values():
+        metric_name = message["name"]
+        tags = message.get("tags", {})
+
+        parsed_strings = {
+            metric_name,
+            *tags.keys(),
+            *tags.values(),
+        }
+        strings.update(parsed_strings)
+
+    with metrics.timer("metrics_consumer.bulk_record"):
+        mapping = indexer.bulk_record(list(strings))
+
+    new_messages: List[Message[KafkaPayload]] = []
+    celery_messages: List[Mapping[str, Union[str, int, Mapping[int, int]]]] = []
+
+    for message in outer_message.payload:
+        parsed_payload_value = parsed_payloads_by_offset[message.offset]
+        new_payload_value = parsed_payload_value
+
+        metric_name = parsed_payload_value["name"]
+        tags = parsed_payload_value.get("tags", {})
+
+        new_tags: Mapping[int, int] = {mapping[k]: mapping[v] for k, v in tags.items()}
+
+        new_payload_value["tags"] = new_tags
+        new_payload_value["metric_id"] = mapping[metric_name]
+        new_payload_value["retention_days"] = 90
+
+        del new_payload_value["name"]
+
+        new_payload = KafkaPayload(
+            key=message.payload.key,
+            value=json.dumps(new_payload_value).encode(),
+            headers=message.payload.headers,
+        )
+        new_message = Message(
+            partition=message.partition,
+            offset=message.offset,
+            payload=new_payload,
+            timestamp=message.timestamp,
+        )
+        new_messages.append(new_message)
+
+        # TODO(meredith): Need to add kickin off the celery task in here, eventually
+        # we will use kafka to forward messages to the product data model
+        celery_messages.append(
+            {"name": metric_name, "tags": new_tags, "org_id": parsed_payload_value["org_id"]}
+        )
+
+    task.apply_async(kwargs={"messages": celery_messages})
+    metrics.incr("metrics_consumer.process_message.messages_seen", amount=len(new_messages))
+
+    return new_messages
+
+
+class MetricsConsumerStrategyFactory(ProcessingStrategyFactory):  # type: ignore
+    def __init__(
+        self,
+        max_batch_size: int,
+        max_batch_time: float,
+        processes: int,
+        input_block_size: int,
+        output_block_size: int,
+    ):
+        self.__max_batch_time = max_batch_time
+        self.__max_batch_size = max_batch_size
+
+        self.__processes = processes
+
+        self.__input_block_size = input_block_size
+        self.__output_block_size = output_block_size
+
+    def create(
+        self, commit: Callable[[Mapping[Partition, Position]], None]
+    ) -> ProcessingStrategy[KafkaPayload]:
+
+        parallel_strategy = ParallelTransformStep(
```

Comment:
We also have metrics to measure how many messages we are batching in the parallelTransformStep (batches of batches). 
https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/streaming/transform.py#L352-L353

Let's keep an eye on these to see how many batches of messages we can fit into a batch to be delivered to a subprocess. We may want to not have too many batches here. But worse, since we are pre-batching data we need to be sure we do not create batches that are too big in bytes. 

In the end we may want to set the batch size of MetricsBatchBuilder in bytes as well so that we ensure that one batch for the MetricsBatchBuilder will never become bigger than the shared memory block, which I guess will cause the whole thing to fail. Watch out for this when rolling out